### ISSUE TITLE
investigate learning rate scheduling

### ISSUE THREAD

**Issue by SteveOv** [SteveOv/ebop_maven#69]

**Comment by SteveOv** [comment]
```Python
# Based on there being 1000 mini-batches per epoch
warmup_steps = 1000
decay_steps = 100000
lr = keras.optimizers.schedules.CosineDecay(0.0, warmup_steps=warmup_steps, warmup_target=5e-4,
                                            decay_steps=decay_steps, alpha=1e-3)
OPTIMIZER = optimizers.Nadam(learning_rate=lr)
```

```
Epoch 106/250
1000/1000 - 81s - 81ms/step - loss: 0.0372 - mse: 0.0070 - val_loss: 0.0316 - val_mse: 0.0069
...
Metrics for 20000 system(s).
----------------------------------------------------------------------------------------------------
Summary    | rA_plus_rB          k          J      ecosw      esinw         bP        MAE        MSE
====================================================================================================
MAE        |   0.006998   0.071812   0.079613   0.024227   0.042422   0.069290   0.049060
MSE        |   0.000202   0.015475   0.065219   0.004485   0.006790   0.014987              0.017860

```

**Comment by SteveOv** [comment]
Baseline with latest split output model, 100k training insts, new synth distributions and Nadam with lr of 5e-4 (which has consistently been the best option from model searches w/o learning rate scheduling).

I've set PATIENCE = 7 for these and we're only using the first 10 files of the 250k training set (so 80k train, 20k valid insts).

```Python
OPTIMIZER = optimizers.Nadam(learning_rate=5e-4)
```

```
Epoch 29/250
1000/1000 - 40s - 40ms/step - loss: 0.0537 - mse: 0.0102 - val_loss: 0.0413 - val_mse: 0.0087
...
Metrics for 20000 system(s).
----------------------------------------------------------------------------------------------------
Summary    | rA_plus_rB          k          J      ecosw      esinw         bP        MAE        MSE
====================================================================================================
MAE        |   0.007682   0.082063   0.095633   0.034666   0.042838   0.082724   0.057601
MSE        |   0.000125   0.016604   0.081780   0.007508   0.006287   0.018697              0.021833
```

**Comment by SteveOv** [comment]
Exponential schedule
```Python
# Based on 1000 steps per training epoch, so should decay by factor ~0.9/epoch
lr_schedule = optimizers.schedules.ExponentialDecay(1e-3, decay_steps=1000, decay_rate=0.9)
OPTIMIZER = optimizers.Nadam(learning_rate=lr_schedule)
```

**Comment by SteveOv** [comment]
Cosine decay with warmup

```Python
# Based on 1000 steps per training epoch, so warmup for 3 epochs then cosine decay for 100
lr_schedule = optimizers.schedules.CosineDecay(initial_learning_rate=0.0, warmup_steps=3000,
                                               warmup_target=1e-3, decay_steps=100000, alpha=0.001)
OPTIMIZER = optimizers.Nadam(learning_rate=lr_schedule)
```

**Comment by SteveOv** [comment]
Cosine decay & warmup with altered params (from initial "what will this do" run).

```Python
# Based on 1000 steps per training epoch, so warmup for 1 epoch then cosine decay for 100
lr_schedule = optimizers.schedules.CosineDecay(initial_learning_rate=0.0, warmup_steps=1000,
                                               warmup_target=5e-4, decay_steps=100000, alpha=1e-3)
OPTIMIZER = optimizers.Nadam(learning_rate=lr_schedule)
```

**Comment by SteveOv** [comment]
OK, there's enough in this for further investigation.  Have set up model_search to include Exponential, Cosine and Piecewise decay in its grid.
